packages/block-library/src/post-tags/index.php

```diff
@@ -20,11 +20,17 @@ function render_block_core_post_tags( $attributes, $content, $block ) {
 
 	$post_tags = get_the_tags( $block->context['postId'] );
 	if ( ! empty( $post_tags ) ) {
-		$output = '';
+		$output  = '';
+		$output .= '<div>';
```

Comment:
Why do we need the additional `<div />` wrapper? :thinking:

Reply:
There are margin styles that target the direct children of entry content for the front-end. Without the wrapping div, the direct children for entry content becomes a series of `<a /> tags`, and spacing margins aren't applied.

<img width="931" alt="Screen Shot 2020-07-23 at 1 15 24 PM" src="https://user-images.githubusercontent.com/5414230/88334795-26483700-cce7-11ea-83c7-f5d330dbdd16.png">
<img width="924" alt="Screen Shot 2020-07-23 at 1 15 14 PM" src="https://user-images.githubusercontent.com/5414230/88334803-29432780-cce7-11ea-9376-b1ba43757e79.png">

Reply:
> There are margin styles that target the direct children of entry content for the front-end. Without the wrapping div, the direct children for entry content become a series of `<a /> tags`, and spacing margins aren't applied. Let me know if that doesn't make sense.

Thanks! This does make sense. Sounds like that's done by the theme though? If it is, it could be argued that it's a theme bug (and should be fixed there rather than here) :thinking: 

(My point is: A different theme could have styling that formats direct `<div />` children of entry content weirdly. So there's no guarantee that what we do to a block's rendered output makes it work for all themes; themes also need to play by some rules.)

Reply:
Since I don't want to block this PR any further, how about we roll back these changes for now and carry them over to another issue or PR for discussion? Arguably, it's not directly related to the goal of this PR (fixing bugs where no tags are rendered).

Reply:
> Sounds like that's done by the theme though?

Ah that's a great point. Digging a little further reveals that the styles originate from the `seedlet` theme.

> If it is, it could be argued that it's a theme bug (and should be fixed there rather than here) 🤔
> A different theme could have styling that formats direct <div /> children of entry content weirdly.

Totally agree. Thanks for the insight @ockham 

> Since I don't want to block this PR any further, how about we roll back these changes for now and carry them over to another issue or PR for discussion?

Sounds good. I'll open an issue on `Automattic/Themes` and roll back the changes now.

Reply:
https://github.com/Automattic/themes/issues/2288

Reply:
Sorry to chime in on this so late but, as far as I can see, _most_ (all?) block-first themes handle the positioning of the content at block level to improve support for different block alignment (wide, full).
In other words, this:
```scss
.entry-content {
  margin: 0 auto;
  width: 700px;
}
```
has been replaced by this:
```scss
.entry-content > * {
  margin: 0 auto;
  width: 700px
}
```

For this reason, if we don't wrap a block inside a `div` (or any block element really, see the Paragraph block using just `p`), it won't be positioned appropriately relative to the content.
Furthermore, we do need a wrapper element to apply the block customization classes and styles.

Reply:
Ah I think @ockham replied [this comment](https://github.com/WordPress/gutenberg/pull/24082#discussion_r460791766) to the wrong thread. 🙂 

> Maybe we should give this block a light block wrapper though, since that seems to be the sort of interface we tend to base some of this customizability on.

Big +1
It's also very easy to implement (once you know what you need to do at least 😄 ).

Reply:
> Ah I think @ockham replied this comment to the wrong thread. slightly_smiling_face

:man_facepalming: Thanks @Copons, I'll move it over to this thread.

Reply:
> Ah I think @ockham replied [this comment](https://github.com/WordPress/gutenberg/pull/24082#discussion_r460791766) to the wrong thread. 
> 
> > Maybe we should give this block a light block wrapper though, since that seems to be the sort of interface we tend to base some of this customizability on.
> 
> Big +1
> It's also very easy to implement (once you know what you need to do at least  ).

On second thought, are light block wrappers available in for dynamic blocks (i.e. server-side, PHP rendered)? :thinking:

Reply:
> On second thought, are light block wrappers available in for dynamic blocks (i.e. server-side, PHP rendered)? 🤔

@ockham As far as I can understand the light block wrapper is supposed to help keep the _editor_ DOM as lean as possible, (among other reasons) in order to look as close as possible like its rendered counterpart, which reduce the differences between editor and front end styles.

So, turning a block into a light block doesn't affect the rendered version (both JS and PHP).
We'll still need to add that `<div>` wrapper in the PHP side of this PR. 🙂

Reply:
> Sorry to chime in on this so late

No worries! 🙂 I always learn something new when you comment.

@Copons Would this `Block.div` component surrounding the `post date` block be an example of a [light block wrapper](https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/post-date/edit.js#L102-L106)? And is there a standardized way to implement them? I had a hard time finding info in MGS.

>So, turning a block into a light block doesn't affect the rendered version (both JS and PHP).

This was a little surprising to me when I first saw it. By rendering information server side and client side, it seems like we're making an unnecessary distinction between blocks created from initial load (PHP) vs. blocks not created from initial load (JS). It doesn't feel like a huge problem, but I'm curious about the reasoning. 

> We'll still need to add that div wrapper in the PHP side of this PR.

If that's the case, then I'm happy to add the div back in 😛

Reply:
> This was a little surprising to me when I first saw it. By rendering information server side and client side, it seems like we're making an unnecessary distinction between blocks created from initial load (PHP) vs. blocks not created from initial load (JS). It doesn't feel like a huge problem, but I'm curious about the reasoning. (The alternative might be serving all data necessary for initial load from an endpoint. JS / React could fetch and render from there)

In fact, it's kind of the other way around! 😄 
Blocks with a JS render (the `save.js` file) save the HTML output of that function straight in the post content.
Blocks with a PHP render instead only save a "token" of that block (e.g. `<!-- core/foo-bar -->` + attributes) in the post content, and replace it with the PHP output when rendering the post in the front end.

The JS render is "static" (relatively speaking): the content of the block can only be edited in the editor.
The PHP render instead can be used to display dynamic content: the content of the block can contain stuff coming from elsewhere. FSE blocks are a good example: you can add the Post Tags block in the editor, but its content comes from an SQL query.

Ideally, blocks should use the JS render as much as possible, in order to persist their output in the post content as normal HTML, to be easily exported and reused.
At the same time, we can't just persist the tags in the post content, because the block then wouldn't pick up any changes on tags.

So basically the JS render doesn't run in the front end, but only when saving the post.
There may be blocks that need some JS in the front end (say, a carousel block), in which case we would need to enqueue its own "front end JS" file.
I'm not sure if Core Gutenberg has a standard way to implement it, but for example many Jetpack blocks have a `view.js` file with that exact purpose.

Reply:
> No worries! 🙂 I always learn something new when you comment.

❤️ 

> @Copons Would this `Block.div` component surrounding the `post date` block be an example of a [light block wrapper](https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/post-date/edit.js#L102-L106)? And is there a standardized way to implement them?

Yes, the Post Date would be a good example.
Basically you need 2 pieces:
- Add support for `lightBlockWrapper `.

https://github.com/WordPress/gutenberg/blob/b67cd9a9ebb621b8711053fcc44f7030c7dfaee4/packages/block-library/src/post-date/block.json#L18

- Output the block as `<Block.div>` (you can choose another tag, e.g. `<Block.p>`, but I don't think it works appropriately with inline elements).

https://github.com/WordPress/gutenberg/blob/b67cd9a9ebb621b8711053fcc44f7030c7dfaee4/packages/block-library/src/post-date/edit.js#L16

https://github.com/WordPress/gutenberg/blob/b67cd9a9ebb621b8711053fcc44f7030c7dfaee4/packages/block-library/src/post-date/edit.js#L102

---

One important (and obvious, in hindsight) thing to keep in mind is that `Block.div` is now the wrapper!

You can keep `BlockControls` and `InspectorControls` outside of it, as they will be automagically rendered where appropriate with the Slot/Fill mechanic.
On the other hand, anything that should appear "inside" the block, must live inside it.

Example: when adding a "post not found" `Warning` in another FSE block, I put it _outside_ the `Block.div` wrapper. What happened was that the warning was rendered in the editor at full width, without any block classes, wrong margins etc.
Here is how I fixed it:
https://github.com/WordPress/gutenberg/blob/b67cd9a9ebb621b8711053fcc44f7030c7dfaee4/packages/block-library/src/post-comments-count/edit.js#L56-L68

Reply:
- Reintroduced the server-side rendered wrapping div
- Added the light block wrapper to markup in `edit.js`

Reply:
I might be wrong on the technicalities, but I think it might be more clear by peeking in the code view (post editor only), where you can see the raw post content.

When you save a post containing a Paragraph block (which has a "JS render"), the rendered HTML of its [`save` function](https://github.com/WordPress/gutenberg/blob/71bf1b5146d5dd0fc15792c1a98fc19df25b062f/packages/block-library/src/paragraph/save.js) is stored in the post content like this:
```html
<!-- wp:paragraph {"textColor":"primary"} -->
<p class="has-primary-color has-text-color">Lorem ipsum</p>
<!-- /wp:paragraph -->
```
You can view this block in the front end without any additional process: it's just plain old HTML, and will work fine anywhere.
Since the "JS render" only runs on post save, we can't use it to fetch dynamic content.

Conversely, when you use a "dynamic" block such as Post Tags, there's no render happening on post save: the editor simply adds the block's "token" (or comment, or whatever we want to call it 🙂) to the post content:
```html
<!-- wp:post-tags {"foo":"bar"} /-->
```
Once in the front end, the block's PHP render callback will basically take that comment, grabs the attributes object, and replace it with the render callback output.
The downside is that this block will only work on WP.

But... blocks can also have both a JS and PHP render! 😄 
Many Jetpack blocks do it in order to have a fallback saved in the post content in case the block is not available anymore (e.g. if the user uninstalls Jetpack).

Maybe a good way to think about it is that blocks are HTML comments.
They can wrap some HTML content (generated on post save by the `save` function), and/or will be replaced by a PHP render function on the front end.
EASY! 😆